### PR TITLE
Add support for extra kotlinc args

### PR DESCRIPTION
**Pull request by cwoodwar6**
This provides the ability to add kotlinc plugins such as kotlin android extensions. Both `robolectric_test` rule and `android_library` rules needed support.

Since `extra_arguments` only applies to the java compiler, we needed a separate argument for the kotlin compiler.

Example of its use (for kotlin android extensions):

```
extra_kotlinc_arguments = [
   '-Xplugin=<kotlin_home>/kotlin-android-extensions.jar',
   '-P',
   'plugin:org.jetbrains.kotlin.android:variant=main;<project-root>/app/src 
    /main/res,plugin:org.jetbrains.kotlin.android:package=com.example.app',
],
```

### CODE REVIEW

File: src/com/facebook/buck/android/AndroidLibraryDescription.java [hunk 1]
```diff
@@ -153,6 +154,8 @@ public void findDepsForTargetFromConstructorArgs(
     Optional<Query> getDepsQuery();
 
     Optional<Query> getProvidedDepsQuery();
+
+    ImmutableList<String> getExtraKotlincArguments();
```
Comment (kageiit) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: We should add this to kotlin library description as well

Comment (cwoodwar6) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: Kotlin library already has `extra_arguments` which pass arguments to the kotlin compiler. Android library is different since it invokes two compilers.

Comment (kageiit) [inline src/com/facebook/buck/android/AndroidLibraryDescription.java]: ah cool

File: src/com/facebook/buck/jvm/java/AndroidCoreArg.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+package com.facebook.buck.jvm.java;
+
+import com.facebook.buck.android.AndroidLibraryDescription;
+import com.google.common.collect.ImmutableList;
+import java.util.Optional;
+
+public interface AndroidCoreArg {
+  Optional<AndroidLibraryDescription.JvmLanguage> getLanguage();
+
+  ImmutableList<String> getExtraKotlincArguments();
```
Comment (kageiit) [inline src/com/facebook/buck/jvm/java/AndroidCoreArg.java]: Can we make this `Optional` instead?

**Comment by cwoodwar6** [comment]
@dsyang - Could you take a look at this one too please?

File: src/com/facebook/buck/jvm/java/AndroidCoreArg.java (hunk 2 above)
Comment (kageiit) [inline src/com/facebook/buck/jvm/java/AndroidCoreArg.java]: @cwoodwar6 not sure if you saw this comment but we should make this optional

Comment (cwoodwar6) [inline src/com/facebook/buck/jvm/java/AndroidCoreArg.java]: Yep, I'll update it.

File: src/com/facebook/buck/jvm/java/AndroidCoreArg.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.facebook.buck.jvm.java;
+
+import com.facebook.buck.android.AndroidLibraryDescription;
+import com.google.common.collect.ImmutableList;
+import java.util.Optional;
+
+public interface AndroidCoreArg {
```
Comment (dsyang) [inline src/com/facebook/buck/jvm/java/AndroidCoreArg.java]: buck.jvm.java seems like the wrong place to put Android stuff. Can this be moved to com.facebook.buck.android?

Comment (cwoodwar6) [inline src/com/facebook/buck/jvm/java/AndroidCoreArg.java]: Yep, I'll move it.

File: src/com/facebook/buck/android/AndroidCoreArg.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.facebook.buck.android;
+
+import com.google.common.collect.ImmutableList;
+import java.util.Optional;
+
+public interface AndroidCoreArg {
```
Comment (dsyang) [inline src/com/facebook/buck/android/AndroidCoreArg.java]: Last small nit: these fields don't seem related to "AndroidCore". `AndroidLibraryDescription.CoreArg` is more "core" than this. Maybe rename to AndroidKotlinCoreArg?

**Comment by cwoodwar6** [comment]
I added a test, but it must intentionally fails until we bump the kotlin version in https://github.com/facebook/buck/issues/1371. Then we can test how to use the plugin appropriately.

File: src/com/facebook/buck/android/AndroidCoreArg.java (hunk 4 above)
Comment (cwoodwar6) [inline src/com/facebook/buck/android/AndroidCoreArg.java]: Good suggestion.

**Comment by cwoodwar6** [comment]
@dsyang - Are we all good here?

**Comment by dsyang** [comment]
@cwoodwar6  I'm having trouble importing this PR. when you get the chance can you rebase?

**Comment by cwoodwar6** [comment]
Just rebased, how about now?
